File: browser/linkedin.js
/* Extractor de LinkedIn. Requiere common.js y una pestaña en linkedin.com (CORS).
 *
 * Flujo típico de una ejecución:
 *   await __radar.li.buscar(TITULOS, {remoto:true, location:'Spain'})
 *   await __radar.li.buscar(TITULOS, {location:'Navarre, Spain'})
 *   __radar.li.filtrar(IDS_CONOCIDOS, '2026-09-08')      // por título y fecha
 *   await __radar.li.detallar(20)                        // en tandas
 *   __radar.li.clasificar(); __radar.saca(6)             // sólo lo que sobrevive
 *
 * Carga también vocabulario.js ANTES de detallar(): desde el 16-sep-2026
 * detallar() saca los términos del vocabulario en la misma pasada que la
 * modalidad, para no volver a pedir la misma ficha dos veces (una para
 * filtrar, otra para escribir los `reqs` del paso 6). Si vocabulario.js no
 * está cargado, detallar() sigue funcionando igual, sólo que sin `j.terminos`
 * — li.terminos() cae entonces a calcularlo sobre `_tn` como hacía antes. */
(function (R) {
  const li = R.li = {};
  li.version = 'linkedin-2026-09-16';
  li.jobs = {};        // id -> ficha del listado
  li.det = [];         // fichas con descripción ya procesada
  li.cola = [];        // pendientes de detallar

  const GUEST = 'https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search?';
  const FICHA = 'https://www.linkedin.com/jobs-guest/jobs/api/jobPosting/';

  /* El HTML del endpoint de invitado NO se puede parsear con DOMParser desde
   * linkedin.com: devuelve 0 nodos. Hay que partir por <li> y sacar los campos
   * con regex sobre el HTML crudo. */
  li.parsear = (html, etiqueta) => {
    let n = 0;
    for (const trozo of html.split('<li')) {
      const id = (trozo.match(/data-entity-urn="urn:li:jobPosting:(\d+)"/) || [])[1];
      if (!id || li.jobs[id]) continue;
      const limpia = s => s ? R.texto(s) : '';
      li.jobs[id] = {
        id,
        titulo: limpia((trozo.match(/<h3[^>]*base-search-card__title[^>]*>([\s\S]*?)<\/h3>/) || [])[1]),
        empresa: limpia((trozo.match(/hidden-nested-link[^>]*>([\s\S]*?)<\/a>/) || [])[1]),
        ubicacion: limpia((trozo.match(/job-search-card__location[^>]*>([\s\S]*?)<\/span>/) || [])[1]),
        fecha: (trozo.match(/datetime="([\d-]+)"/) || [])[1] || '',
        q: etiqueta,
      };
      n++;
    }
    return n;
  };

  /* opts: {remoto, location, horas (por defecto 24), paginas} */
  li.buscar = async (titulos, opts) => {
    opts = opts || {};
    const horas = opts.horas || 24;
    const paginas = opts.paginas || 2;
    const consultas = [];
    for (const t of titulos)
      for (let p = 0; p < paginas; p++)
        consultas.push({
          etiqueta: (opts.remoto ? 'R' : 'L') + (p ? '2' : '') + '|' + t,
          qs: 'keywords=' + encodeURIComponent(t)
            + '&location=' + encodeURIComponent(opts.location || 'Spain')
            + '&f_TPR=r' + (horas * 3600)
            + (opts.remoto ? '&f_WT=2' : '')
            + '&start=' + (p * 10),
        });
    const res = await R.enTandas(consultas, async c => {
      const r = await R.pide(GUEST + c.qs);
      return c.etiqueta + ':' + r.status + ':+' + li.parsear(await r.text(), c.etiqueta);
    }, 220);
    return res.join(' ') + ' TOTAL=' + Object.keys(li.jobs).length;
  };

  /* Criba por título, por fecha y contra los ids que ya están en la base.
   * Es lo que baja de ~150 candidatas a ~20 fichas que merezca la pena pedir. */
  li.filtrar = (idsConocidos, desde, excluirEmpresas) => {
    const conocidos = new Set((idsConocidos || []).map(String));
    const malas = (excluirEmpresas || []).map(R.norm).filter(Boolean);
    li.cola = Object.values(li.jobs).filter(j => {
      if (conocidos.has(j.id)) return false;
      if (desde && j.fecha && j.fecha < desde) return false;
      if (!R.tituloVale(j.titulo)) return false;
      const e = R.norm(j.empresa);
      if (malas.some(m => e === m || (m.length > 4 && e.includes(m)))) return false;
      return true;
    });
    return { candidatas: Object.keys(li.jobs).length, sobreviven: li.cola.length };
  };

  /* Calcula los hits de vocabulario sobre texto ya normalizado. Compartida
   * con terminos() para no repetir la cuenta si ya está en j.terminos.
   *
   * Cada hit lleva un fragmento corto (~110 caracteres) alrededor de la
   * PRIMERA aparición del término: no para contar, sino para que al escribir
   * `reqs` en el paso 6 se pueda poner el peso («imprescindible» pesa más que
   * «se valorará») sin leer la ficha entera — ver snippets() más abajo, y el
   * porqué en la cabecera de vocabulario.js. */
  const _cuentaTerminos = (tn, dicc) => {
    const D = dicc || R.DICC;
    if (!D || !tn) return null;
    const hits = [];
    for (const k in D) {
      const re = D[k];
      const m = tn.match(new RegExp(re, 'g'));
      if (!m) continue;
      const idx = tn.search(new RegExp(re));
      const frag = idx >= 0 ? tn.slice(Math.max(0, idx - 40), idx + 70).replace(/\s+/g, ' ').trim() : '';
      hits.push([k, m.length, frag]);
    }
    hits.sort((a, b) => b[1] - a[1]);
    return hits;
  };

  /* Descarga la descripción y saca sólo los campos que importan: modalidad,
   * ámbito, salario, años y —si vocabulario.js está cargado— los términos
   * del vocabulario, todo de la misma lectura. Nunca get_page_text ni
   * read_page sobre una ficha: son miles de tokens. */
  li.detallar = async (n) => {
    const lote = li.cola.splice(0, n || 20);
    await R.enTandas(lote, async j => {
      const r = await R.pide(FICHA + j.id);
      if (r.status === 404) { j.cerrada = true; li.det.push(j); return; }
      const txt = R.texto(await r.text());
      const tn = R.norm(txt);
      const etiquetaRemoto = /^R/.test(j.q || '');
      Object.assign(j, {
        modalidad: R.modalidad(tn, R.norm(j.ubicacion), etiquetaRemoto),
        ambito: R.ambito(tn),
        salario: R.salario(txt),
        anios: R.anios(tn),
        largo: txt.length,
        terminos: _cuentaTerminos(tn),   // null si no hay vocabulario.js cargado
      });
      j._tn = tn;   // se guarda para leer() y para terminos() si hace falta recalcular
      li.det.push(j);
    }, 230);
    return 'detalladas=' + li.det.length + ' pendientes=' + li.cola.length;
  };

  /* Reparte en las tres cestas y deja las buenas en la cola de salida. */
  li.clasificar = (areasLocales) => {
    const dentro = [], fuera = [], revisar = [];
    for (const j of li.det) {
      if (j.cerrada) { fuera.push(j); continue; }
      const t = j.modalidad.tipo;
      if (t === 'remoto' || t === 'local') dentro.push(j);
      else if (t === 'remoto_sin_confirmar') revisar.push(j);
      else fuera.push(j);
    }
    const fila = j => [j.id, j.empresa, j.titulo, j.ubicacion, j.fecha,
                       j.modalidad.tipo, j.salario || '-', j.anios || '-',
                       (j.ambito.restriccion || '').slice(0, 60),
                       (j.modalidad.frases[0] || '').slice(0, 90)].join('|');
    R.cola = dentro.concat(revisar).map(fila);
    return { dentro: dentro.length, revisar: revisar.length, fuera: fuera.length };
  };

  /* Descripción completa de una oferta ya aceptada. Antes era la única forma
   * de escribir sus `reqs`; desde el 16-sep-2026 usa snippets() en su lugar
   * para eso (ver más abajo) y esto queda como último recurso, para cuando
   * la lista de términos salga corta o rara, o para redactar `resumen`.
   * Sale a trozos porque la salida se corta sobre los 1.200 caracteres. */
  li.leer = (id, desde, chars) => {
    const j = li.det.find(x => x.id === String(id));
    if (!j || !j._tn) return 'no encontrada (¿detallada?)';
    return j._tn.slice(desde || 0, (desde || 0) + (chars || 900));
  };

  /* Alternativa barata: sólo el vocabulario técnico, contado sobre la
   * descripción. Usa lo que detallar() ya calculó (j.terminos); si esa
   * oferta se detalló antes de cargar vocabulario.js, lo calcula ahora
   * sobre _tn con el diccionario que se le pase (o R.DICC). */
  li.terminos = (ids, dicc) => li.det
    .filter(j => !ids || ids.includes(j.id))
    .map(j => {
      const hits = j.terminos || _cuentaTerminos(j._tn, dicc) || [];
      return j.id + '|' + hits.slice(0, 22).map(h => h[0] + ':' + h[1]).join(',');
    });

  /* Lo que hace falta para escribir los `reqs` del paso 6 SIN leer la ficha
   * entera: los `maxTerms` términos con más apariciones, cada uno con un
   * trozo corto de dónde aparece por primera vez, para poner el peso y la
   * etiqueta con la frase real del anuncio. Una sola oferta por llamada
   * (las 4-5 que sobreviven al día), así que cabe de sobra en los ~1.200
   * caracteres de salida. Si sale vacío o insuficiente, cae a li.leer(). */
  li.snippets = (id, maxTerms) => {
    const j = li.det.find(x => x.id === String(id));
    if (!j) return 'no encontrada (¿detallada?)';
    const hits = j.terminos || _cuentaTerminos(j._tn) || [];
    if (!hits.length) return '(sin términos de vocabulario.js — usa li.leer())';
    return hits.slice(0, maxTerms || 10)
      .map(h => `${h[0]} (${h[1]}x): "${h[2] || ''}"`)
      .join('\n');
  };
})(window.__radar);

File: pipeline/fuentes/linkedin.py
# -*- coding: utf-8 -*-
"""Extractor de LinkedIn — portado de `browser/linkedin.js` a Python. Desde el
21-sep-2026 el HTML lo trae Scrapling (`ScraplingServer.fetch`/`bulk_fetch`,
`real_chrome:true`), no una pestaña de linkedin.com — así que ya no hace
falta pegar código ni estar en el propio dominio (era sólo para saltarse la
CSP y el CORS). Este script sólo analiza lo que Scrapling ya descargó.

Flujo típico de una ejecución (ver TAREA_DIARIA.md, Pasos 2-4):

  1. `consultas`   -> URLs a pedir con Scrapling (el agente las pasa a
                      `bulk_fetch`, extraction_type "html", real_chrome:true).
  2. `parsear`     -> junta el HTML de esas páginas en un diccionario de
                      ofertas (id, título, empresa, ubicación, fecha).
  3. `filtrar`     -> criba por título, fecha y empresas excluidas: de ~150
                      candidatas a ~20 fichas que merece la pena pedir.
  4. (Scrapling pide la ficha de cada superviviente: extraction_type "html")
  5. `detallar`    -> modalidad, ámbito, salario, años y términos de
                      vocabulario, todo de la misma lectura.
  6. `clasificar`  -> reparte en dentro/revisar/fuera según
                      `buscar_remoto`/`buscar_hibrido`/`buscar_presencial`.
  7. `snippets`/`leer` -> para redactar los `reqs` y el `resumen` del paso 6.

El endpoint de invitado (`jobs-guest/.../search`) NO se puede parsear con un
parser de DOM normal: hay que partir por `<li` y sacar los campos con regex
sobre el HTML crudo — eso no ha cambiado, es una característica del propio
HTML de LinkedIn, no del método de descarga.
"""
import argparse
import json
import os
import re
import sys

# Permite `python pipeline/fuentes/linkedin.py ...` desde la raíz del repo
# (la convención de todo el pipeline) sin instalar el paquete.
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
from pipeline.fuentes import comun, vocabulario

GUEST = "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search?"
FICHA = "https://www.linkedin.com/jobs-guest/jobs/api/jobPosting/"
FICHA_COMPLETA = "https://www.linkedin.com/jobs/view/"

_RE_ID = re.compile(r'data-entity-urn="urn:li:jobPosting:(\d+)"')
_RE_TITULO = re.compile(r'<h3[^>]*base-search-card__title[^>]*>([\s\S]*?)</h3>')
_RE_EMPRESA = re.compile(r'hidden-nested-link[^>]*>([\s\S]*?)</a>')
_RE_UBICACION = re.compile(r'job-search-card__location[^>]*>([\s\S]*?)</span>')
_RE_FECHA = re.compile(r'datetime="([\d-]+)"')

# Ver linkedin.js: la insignia («Remoto»/«Híbrido»/«Presencial») sólo vive en
# la página completa (jobs/view/<id>), no en la ficha ligera del endpoint de
# invitado, y sólo se pide para las pocas ofertas que quedan en
# `remoto_sin_confirmar` tras leer la descripción — no para todas.
_RE_ETIQUETA = re.compile(r'>\s*(Remoto|H[ií]brido|Presencial|Remote|Hybrid|On-?site)\s*<')
_MAPA_ETIQUETA = {
    "remoto": "remoto", "remote": "remoto",
    "hibrido": "hibrido", "hybrid": "hibrido",
    "presencial": "presencial", "on-site": "presencial", "onsite": "presencial",
}


def construir_consultas(titulos, location="Spain", horas=24, paginas=2, remoto=False):
    """Igual que `li.buscar()` en construir las URLs, sin hacer la petición:
    eso lo hace Scrapling. `paginas` = nº de páginas de 10 resultados por
    título (start=0,10,20...)."""
    out = []
    for t in titulos:
        for p in range(paginas):
            etiqueta = ("R" if remoto else "L") + ("2" if p else "") + "|" + t
            qs = {
                "keywords": t,
                "location": location,
                "f_TPR": "r" + str(horas * 3600),
            }
            if remoto:
                qs["f_WT"] = "2"
            qs["start"] = str(p * 10)
            from urllib.parse import urlencode
            out.append({"etiqueta": etiqueta, "url": GUEST + urlencode(qs)})
    return out


def parsear(html, etiqueta, jobs=None):
    """Trocea por `<li` y saca los campos con regex sobre el HTML crudo (el
    endpoint de invitado no se puede parsear con DOMParser/BeautifulSoup de
    forma fiable). Devuelve (jobs actualizado, nº de ofertas nuevas)."""
    jobs = jobs if jobs is not None else {}
    n = 0
    for trozo in html.split("<li"):
        m_id = _RE_ID.search(trozo)
        if not m_id:
            continue
        jid = m_id.group(1)
        if jid in jobs:
            continue
        limpia = lambda mm: comun.texto(mm.group(1)) if mm else ""
        jobs[jid] = {
            "id": jid,
            "titulo": limpia(_RE_TITULO.search(trozo)),
            "empresa": limpia(_RE_EMPRESA.search(trozo)),
            "ubicacion": limpia(_RE_UBICACION.search(trozo)),
            "fecha": (_RE_FECHA.search(trozo).group(1) if _RE_FECHA.search(trozo) else ""),
            "q": etiqueta,
        }
        n += 1
    return jobs, n


def filtrar(jobs, conocidos=None, desde=None, excluir_empresas=None):
    """Criba por título, por fecha y contra los ids ya conocidos. Es lo que
    baja de ~150 candidatas a ~20 fichas que merezca la pena pedir."""
    conocidos = set(str(x) for x in (conocidos or []))
    malas = [comun.norm(e) for e in (excluir_empresas or []) if e]
    cola = []
    for j in jobs.values():
        if j["id"] in conocidos:
            continue
        if desde and j.get("fecha") and j["fecha"] < desde:
            continue
        if not comun.titulo_vale(j["titulo"]):
            continue
        e = comun.norm(j["empresa"])
        if any(e == m or (len(m) > 4 and m in e) for m in malas):
            continue
        cola.append(j)
    return cola


def detallar_una(job, html_ficha):
    """Ficha ligera del endpoint de invitado -> campos calculados. `job` es
    la entrada del listado (para ubicacion/q); `html_ficha` es lo que
    Scrapling trajo de `FICHA + id`."""
    txt = comun.texto(html_ficha)
    tn = comun.norm(txt)
    etiqueta_remoto = job.get("q", "").startswith("R")
    mod = comun.modalidad(tn, comun.norm(job.get("ubicacion", "")), etiqueta_remoto)
    out = dict(job)
    out.update({
        "modalidad": mod,
        "ambito": comun.ambito(tn),
        "salario": comun.salario(txt),
        "anios": comun.anios(tn),
        "largo": len(txt),
        "terminos": vocabulario.cuenta_terminos(tn),
        "_tn": tn,
    })
    return out


def resolver_etiqueta(html_pagina_completa):
    """Lee la insignia de modalidad de la página completa (`jobs/view/<id>`),
    sólo para desambiguar `remoto_sin_confirmar`. Se lee de los primeros
    60.000 caracteres (zona de cabecera, antes de "empleos similares", que
    repite la palabra para OTRAS ofertas)."""
    cabecera = (html_pagina_completa or "")[:60000]
    m = _RE_ETIQUETA.search(cabecera)
    if not m:
        return None
    return _MAPA_ETIQUETA.get(comun.norm(m.group(1)))


def clasificar(detalladas, cfg=None):
    aceptadas = comun.modalidades_aceptadas(cfg)
    dentro, fuera, revisar = [], [], []
    for j in detalladas:
        if j.get("cerrada"):
            fuera.append(j)
            continue
        t = j["modalidad"]["tipo"]
        if t not in aceptadas:
            fuera.append(j)
            continue
        (revisar if t == "remoto_sin_confirmar" else dentro).append(j)

    def fila(j):
        return "|".join([
            j["id"], j["empresa"], j["titulo"], j["ubicacion"], j.get("fecha", ""),
            j["modalidad"]["tipo"], j.get("salario") or "-", j.get("anios") or "-",
            (j.get("ambito", {}).get("restriccion") or "")[:60],
            (j["modalidad"]["frases"][0] if j["modalidad"]["frases"] else "")[:90],
        ])

    filas = [fila(j) for j in dentro + revisar]
    return {"filas": filas, "dentro": len(dentro), "revisar": len(revisar), "fuera": len(fuera)}


def snippets(job, max_terms=10):
    hits = job.get("terminos") or vocabulario.cuenta_terminos(job.get("_tn", "")) or []
    if not hits:
        return "(sin términos de vocabulario — usar leer())"
    return "\n".join(f'{h[0]} ({h[1]}x): "{h[2]}"' for h in hits[:max_terms])


def leer(job, desde=0, chars=900):
    tn = job.get("_tn", "")
    if not tn:
        return "no encontrada (¿detallada?)"
    return tn[desde:desde + chars]


# ---------------------------------------------------------------- CLI ----
def _leer_json(path):
    with open(path, encoding="utf-8") as f:
        return json.load(f)


def _escribir_json(path, obj):
    if path:
        with open(path, "w", encoding="utf-8") as f:
            json.dump(obj, f, ensure_ascii=False, indent=2)
    else:
        print(json.dumps(obj, ensure_ascii=False, indent=2))


def main():
    p = argparse.ArgumentParser(description=__doc__)
    sub = p.add_subparsers(dest="cmd", required=True)

    sp = sub.add_parser("consultas")
    sp.add_argument("--titulos", required=True, help="JSON: lista de títulos")
    sp.add_argument("--location", default="Spain")
    sp.add_argument("--horas", type=int, default=24)
    sp.add_argument("--paginas", type=int, default=2)
    sp.add_argument("--remoto", action="store_true")
    sp.add_argument("--out")

    sp = sub.add_parser("parsear")
    sp.add_argument("--in", dest="entrada", required=True,
                     help='JSON: [{"etiqueta":..., "html":...}, ...]')
    sp.add_argument("--out")

    sp = sub.add_parser("filtrar")
    sp.add_argument("--jobs", required=True)
    sp.add_argument("--conocidos", help="JSON: lista de ids")
    sp.add_argument("--desde")
    sp.add_argument("--excluir-empresas", help="JSON: lista de nombres")
    sp.add_argument("--out")

    sp = sub.add_parser("detallar")
    sp.add_argument("--in", dest="entrada", required=True,
                     help='JSON: [{"id":..., "html":...}, ...] (fichas)')
    sp.add_argument("--jobs", required=True, help="jobs.json del paso parsear")
    sp.add_argument("--out")

    sp = sub.add_parser("resolver-etiquetas")
    sp.add_argument("--in", dest="entrada", required=True,
                     help="JSON de detallar, a corregir in place")
    sp.add_argument("--paginas", required=True,
                     help='JSON: [{"id":..., "html":...}, ...] (jobs/view completas)')
    sp.add_argument("--out")

    sp = sub.add_parser("clasificar")
    sp.add_argument("--in", dest="entrada", required=True)
    sp.add_argument("--config", help="config/filtros como JSON")
    sp.add_argument("--out")

    sp = sub.add_parser("snippets")
    sp.add_argument("--in", dest="entrada", required=True)
    sp.add_argument("--id", required=True)
    sp.add_argument("--max-terms", type=int, default=10)

    sp = sub.add_parser("leer")
    sp.add_argument("--in", dest="entrada", required=True)
    sp.add_argument("--id", required=True)
    sp.add_argument("--desde", type=int, default=0)
    sp.add_argument("--chars", type=int, default=900)

    args = p.parse_args()

    if args.cmd == "consultas":
        titulos = _leer_json(args.titulos)
        out = construir_consultas(titulos, args.location, args.horas, args.paginas, args.remoto)
        _escribir_json(args.out, out)

    elif args.cmd == "parsear":
        entrada = _leer_json(args.entrada)
        jobs = {}
        total_nuevas = 0
        for pagina in entrada:
            jobs, n = parsear(pagina["html"], pagina["etiqueta"], jobs)
            total_nuevas += n
        _escribir_json(args.out, jobs)
        print(f"TOTAL={len(jobs)}", file=sys.stderr)

    elif args.cmd == "filtrar":
        jobs = _leer_json(args.jobs)
        conocidos = _leer_json(args.conocidos) if args.conocidos else []
        excluir = _leer_json(args.excluir_empresas) if args.excluir_empresas else []
        cola = filtrar(jobs, conocidos, args.desde, excluir)
        _escribir_json(args.out, cola)
        print(f"candidatas={len(jobs)} sobreviven={len(cola)}", file=sys.stderr)

    elif args.cmd == "detallar":
        fichas = _leer_json(args.entrada)
        jobs = _leer_json(args.jobs)
        out = []
        for f in fichas:
            job = jobs.get(f["id"], {"id": f["id"], "titulo": "", "empresa": "",
                                      "ubicacion": "", "fecha": "", "q": ""})
            out.append(detallar_una(job, f["html"]))
        _escribir_json(args.out, out)
        print(f"detalladas={len(out)}", file=sys.stderr)

    elif args.cmd == "resolver-etiquetas":
        detalladas = _leer_json(args.entrada)
        paginas = {x["id"]: x["html"] for x in _leer_json(args.paginas)}
        for j in detalladas:
            if j["modalidad"]["tipo"] != "remoto_sin_confirmar":
                continue
            html = paginas.get(j["id"])
            if not html:
                continue
            et = resolver_etiqueta(html)
            if et:
                j["modalidad"]["frases"] = [f"etiqueta de LinkedIn: {et}"] + j["modalidad"]["frases"]
                j["modalidad"]["tipo"] = et
        _escribir_json(args.out, detalladas)

    elif args.cmd == "clasificar":
        detalladas = _leer_json(args.entrada)
        cfg = _leer_json(args.config) if args.config else None
        res = clasificar(detalladas, cfg)
        if args.out:
            _escribir_json(args.out, res["filas"])
        else:
            print("\n".join(res["filas"]))
        print(f"dentro={res['dentro']} revisar={res['revisar']} fuera={res['fuera']}",
              file=sys.stderr)

    elif args.cmd == "snippets":
        detalladas = _leer_json(args.entrada)
        job = next((j for j in detalladas if j["id"] == args.id), None)
        print("no encontrada" if job is None else snippets(job, args.max_terms))

    elif args.cmd == "leer":
        detalladas = _leer_json(args.entrada)
        job = next((j for j in detalladas if j["id"] == args.id), None)
        print("no encontrada" if job is None else leer(job, args.desde, args.chars))


if __name__ == "__main__":
    main()
